python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        """Makes a column optional.
 
-        exist_field: Optional[NestedField] = None
-        if parent:
-            parent_field = self._schema.find_field(parent, self._case_sensitive)
-            parent_type = parent_field.field_type
-            if isinstance(parent_type, MapType):
-                parent_field = parent_type.value_field
-            elif isinstance(parent_type, ListType):
-                parent_field = parent_type.element_field
+        Args:
+            path: The path to the field.
 
-            if not parent_field.field_type.is_struct:
-                raise ValueError(f"Cannot add column to non-struct type: {parent}")
+        Returns:
+            The UpdateSchema with the requirement change staged.
+        """
+        self._set_column_requirement(path, False)
+        return self
 
-            parent_id = parent_field.field_id
+    def _set_column_requirement(self, path: Union[str, Tuple[str, ...]], required: bool) -> None:
+        path = (path,) if isinstance(path, str) else path
+        name = ".".join(path)
+
+        field = self._schema.find_field(name, self._case_sensitive)
 
-            try:
-                exist_field = self._schema.find_field(parent + "." + name, self._case_sensitive)
-            except ValueError:
-                pass
+        if (field.required and required) or (field.optional and not required):
+            # if the change is a noop, allow it even if allowIncompatibleChanges is false
+            return
 
-            if exist_field:
-                raise ValueError(f"Cannot add column, name already exists: {parent}.{name}")
+        if self._allow_incompatible_changes and not required:
+            raise ValueError(f"Cannot change column nullability: {name}: optional -> required")
 
-            full_name = parent_field.name + "." + name
+        if field.field_id in self._deletes:
+            raise ValueError(f"Cannot update a column that will be deleted: {name}")
 
+        if updated := self._updates.get(field.field_id):
+            self._updates[field.field_id] = NestedField(
+                field_id=updated.field_id,
+                name=updated.name,
+                field_type=updated.field_type,
+                doc=updated.doc,
+                required=required,
+            )
         else:
-            try:
-                exist_field = self._schema.find_field(name, self._case_sensitive)
-            except ValueError:
-                pass
+            self._updates[field.field_id] = NestedField(
+                field_id=field.field_id,
+                name=field.name,
+                field_type=field.field_type,
+                doc=field.doc,
+                required=required,
+            )
 
-            if exist_field:
-                raise ValueError(f"Cannot add column, name already exists: {name}")
+    def update_column(self, path: Union[str, Tuple[str, ...]], field_type: IcebergType) -> UpdateSchema:
```

Comment:
Rather than having separate methods like the Java implementation, what about using a combined `update_column` with named parameters?

```python
def update_column(self, path: Union[str, Tuple[str, ...]], field_type: Optional[IcebergType] = None, required: Optional[boolean] = None, doc: Optional[str] = None) -> UpdateSchema:
     # find the column ...
     field_type = field_type if field_type is None else field.field_type
     doc = doc if doc is None else field.doc
     required = required if required is None else field.required
     ...
```

Reply:
I like it, updated!

Reply:
I left the old methods in place, WDYT?